pkg/frontend/querymiddleware/astmapper/sharding.go

```diff
@@ -18,39 +18,67 @@ import (
 )
 
 // NewSharding creates a new query sharding mapper.
-func NewSharding(ctx context.Context, shards int, logger log.Logger, stats *MapperStats) (ASTMapper, error) {
-	shardSummer, err := newShardSummer(ctx, shards, vectorSquasher, logger, stats)
-	if err != nil {
-		return nil, err
-	}
+func NewSharding(shardSummer ASTMapper) ASTMapper {
 	subtreeFolder := newSubtreeFolder()
 	return NewMultiMapper(
 		shardSummer,
 		subtreeFolder,
-	), nil
+	)
+}
+
+type Squasher = func(...parser.Expr) (parser.Expr, error)
+
+type ShardLabeller interface {
+	GetLabelName() string
+	GetLabelValue(shard int) string
 }
 
-type squasher = func(...parser.Expr) (parser.Expr, error)
+// queryShardLabeller implements ShardLabeller for query sharding.
+type queryShardLabeller struct {
+	shards int
+}
+
+func newQueryShardLabeller(shards int) ShardLabeller {
+	return &queryShardLabeller{shards: shards}
+}
+
+func (lbl *queryShardLabeller) GetLabelName() string {
+	return sharding.ShardLabel
+}
+
+func (lbl *queryShardLabeller) GetLabelValue(shard int) string {
+	return sharding.ShardSelector{ShardIndex: uint64(shard), ShardCount: uint64(lbl.shards)}.LabelValue()
+}
+
+// NewQueryShardSummer instantiates an ASTMapper which will fan out sum queries by shard.
+func NewQueryShardSummer(ctx context.Context, shards int, squasher Squasher, logger log.Logger, stats *MapperStats) (ASTMapper, error) {
+	summer, err := NewShardSummer(ctx, shards, squasher, logger, stats, newQueryShardLabeller(shards))
+	if err != nil {
+		return nil, err
+	}
+	return NewASTExprMapper(summer), nil
+}
 
 type shardSummer struct {
 	ctx context.Context
 
 	shards       int
 	currentShard *int
-	squash       squasher
+	squash       Squasher
 	logger       log.Logger
 	stats        *MapperStats
 
+	shardLabeller ShardLabeller
+
 	canShardAllVectorSelectorsCache map[string]bool
 }
 
-// newShardSummer instantiates an ASTMapper which will fan out sum queries by shard
-func newShardSummer(ctx context.Context, shards int, squasher squasher, logger log.Logger, stats *MapperStats) (ASTMapper, error) {
+func NewShardSummer(ctx context.Context, shards int, squasher Squasher, logger log.Logger, stats *MapperStats, shardLabeller ShardLabeller) (*shardSummer, error) { //nolint:revive // We do not need shardSummer to be exported.
```

Comment:
Is this called from anywhere except line 55 above? It seems weird that this method is exported but the type it returns is not. If the only place that calls this method is line 55 above, then can we make this method unexported too?

Reply:
In Mimir it's only line 55, but we're exporting it so that GEM can re-use this for the cluster sharding.

Reply:
Right, makes sense, thanks for explaining.

Reply:
I think in this case, it makes sense to export `shardSummer` then, no?